packages/beacon-node/src/chain/blocks/blockInput/blockInput.ts

```diff
@@ -0,0 +1,786 @@
+    // }
+
+    this.blobsCache.set(blobSidecar.index, {blobSidecar, source, seenTimestampSec, peerIdStr});
+
+    if (this.state.hasBlock && this.blobsCache.size === this.state.block.message.body.blobKzgCommitments.length) {
+      this.state = {
+        ...this.state,
+        hasAllData: true,
+        timeCompleteSec: seenTimestampSec,
+      };
+      this.dataPromise.resolve([...this.blobsCache.values()].map(({blobSidecar}) => blobSidecar));
+    }
+  }
+
+  getMissingBlobMeta(): BlobMeta[] {
+    if (!this.state.hasBlock) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.INCOMPLETE_DATA,
+          ...this.getLogMeta(),
+        },
+        "Cannot get missing blobs. Block is unknown"
+      );
+    }
+    if (this.state.hasAllData) {
+      return [];
+    }
+
+    const blobMeta: BlobMeta[] = [];
+    const versionHashes = this.state.versionHashes;
+    for (let index = 0; index < versionHashes.length; index++) {
+      if (!this.blobsCache.has(index)) {
+        blobMeta.push({
+          index,
+          blockRoot: fromHex(this.blockRootHex),
+          versionHash: versionHashes[index],
+        });
+      }
+    }
+    return blobMeta;
+  }
+
+  getAllBlobsWithSource(): BlobWithSource[] {
+    if (!this.state.hasAllData) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.INCOMPLETE_DATA,
+          ...this.getLogMeta(),
+        },
+        "Cannot get all blobs. DA status is not complete"
+      );
+    }
+    return [...this.blobsCache.values()];
+  }
+
+  getBlobs(): deneb.BlobSidecars {
+    return this.getAllBlobsWithSource().map(({blobSidecar}) => blobSidecar);
+  }
+}
+
+function getVersionHashes(block: SignedBeaconBlock<ForkPostDeneb>): VersionedHashes {
+  return block.message.body.blobKzgCommitments.map(kzgCommitmentToVersionedHash);
+}
+
+function blockAndBlobArePaired(block: SignedBeaconBlock<ForkBlobsDA>, blobSidecar: deneb.BlobSidecar): boolean {
+  return byteArrayEquals(block.message.body.blobKzgCommitments[blobSidecar.index], blobSidecar.kzgCommitment);
+}
+
+function assertBlockAndBlobArePaired(
+  blockRootHex: string,
+  block: SignedBeaconBlock<ForkBlobsDA>,
+  blobSidecar: deneb.BlobSidecar
+): void {
+  if (!blockAndBlobArePaired(block, blobSidecar)) {
+    // TODO: (@matthewkeil) should this eject the bad blob instead? No way to tell if the blob or the block
+    //       has the invalid commitment. Guessing it would be the blob though because we match via block
+    //       hashTreeRoot and we do not take a hashTreeRoot of the BlobSidecar
+    throw new BlockInputError(
+      {
+        code: BlockInputErrorCode.MISMATCHED_KZG_COMMITMENT,
+        blockRoot: blockRootHex,
+        slot: block.message.slot,
+        sidecarIndex: blobSidecar.index,
+      },
+      "BlobSidecar commitment does not match block commitment"
+    );
+  }
+}
+
+// Columns DA
+
+export type ForkColumnsDA = ForkName.fulu;
+
+type BlockInputColumnsState =
```

Comment:
`versionedHashes` will exist and be required for all cases

Reply:
ah! yes column sidecar includes a list of the kzg commitments